broker-core/src/main/java/io/zeebe/broker/subscription/message/processor/PublishMessageProcessor.java

```diff
@@ -74,35 +75,59 @@ public void processRecord(
       responseWriter.writeRejectionOnCommand(command, RejectionType.BAD_VALUE, rejectionReason);
 
     } else {
-      final TypedBatchWriter batchWriter = streamWriter.newBatch();
-      final long key = batchWriter.addNewEvent(MessageIntent.PUBLISHED, command.getValue());
-      responseWriter.writeEventOnCommand(key, MessageIntent.PUBLISHED, command.getValue(), command);
+      handleNewMessage(command, responseWriter, streamWriter, sideEffect);
+    }
+  }
 
-      matchingSubscriptions =
-          messageStateController.findSubscriptions(
-              messageRecord.getName(), messageRecord.getCorrelationKey());
+  private void handleNewMessage(
+      final TypedRecord<MessageRecord> command,
+      final TypedResponseWriter responseWriter,
+      final TypedStreamWriter streamWriter,
+      final Consumer<SideEffectProducer> sideEffect) {
 
-      for (final MessageSubscription sub : matchingSubscriptions) {
-        sub.setMessagePayload(messageRecord.getPayload());
-      }
+    final TypedBatchWriter batchWriter = streamWriter.newBatch();
+    final long key = batchWriter.addNewEvent(MessageIntent.PUBLISHED, command.getValue());
+    responseWriter.writeEventOnCommand(key, MessageIntent.PUBLISHED, command.getValue(), command);
+
+    // correlate the message only once per workflow instance
+    // - will be improved by #1421
+    matchingSubscriptions =
+        messageStateController
+            .findSubscriptions(messageRecord.getName(), messageRecord.getCorrelationKey())
```

Comment:
stream api really?

Reply:
Yes, for now. But I promise that I will remove it soon ;)